### ISSUE TITLE
HLSL: Is it possible to pass unused types into SPIR-V bytecode

### ISSUE THREAD

**Issue by Joojooo** [KhronosGroup/glslang#616]
Hi. Is it possible to have a type defined in the shader source-code put into the SPIR-V bytecode (and through the AST), even though this type is not used by any variables?

For example this struct will not be defined in the resulting SPIR-V, unless I declare a variable using it
static struct aStruct
{
    float4 param;
};

Is it possible to force its definition to be written in the bytecode anyway, either by defining it somehow in the shader file, or by updating the parser source code (I experimented forcefully calling parseContext.trackLinkageDeferred on the new type (after its creation in HlslGrammar::acceptStruct function) but without success) ?

Thanks

**Comment by johnkslang** [comment]
You could try declaring an unused interface variable of that type.

**Comment by Joojooo** [comment]
Yes that's one possibility. But is it also possible to do it without declaring unused variables? Problem is that we don't have control over the HLSL files we're parsing: we receive customs HLSL shaders and want to convert them to SPIR-V while keeping all user-defined types (reasons is that later on we will merge several SPIR-V files together and remove unused types at that moment). Is there possibility to force those types to be written by updating the source ?

**Comment by johnkslang** [comment]
What bad thing happens if an unused type is not passed on to SPIR-V?

Do you have a SPIR-V module-merger project underway?

**Comment by Joojooo** [comment]
Not yet, but we plan to make a SPIR-V files merger later.

I checked how to add an unused type from the parser (I don't want to modify the original shader file)
It works if I add this code at the end of acceptDeclaration function:

`if (declaredType.getBasicType() == EbtStruct) { 

TString unusedVariableName = "__var_" + declaredType.getTypeName() + "_unused"; 
parseContext.declareVariable(initialToken.loc, unusedVariableName, declaredType, nullptr); 
}`

Weird things is that the linker will not use the specified variable name, but apparently the name from the token location (that's why I put initialToken.loc, with the current token (position at end of file) it would make the app to crash.

**Comment by johnkslang** [comment]
Normally, dead types would not be wanted. If you merged, and one module did use the type, you would end up with the type in the final module. So, I still don't understand the use model of wanting dead types in the SPIR-V module. Knowing the purpose might help answer the question.

**Comment by ratchetfreak** [comment]
One use-case I can see is a poor-man's linker, 

Where you just append the function called to a partial module. It'll save the runtime cost of analyzing the types and culling the duplicates along with the other costs for the linking semantics.

**Comment by johnkslang** [comment]
Each module has to bring along types it needs, but weren't in other modules. Once that happens, all needed types will be present. Sorry, I simply don't see why you want an unused type in the SPIR-V module.

**Comment by ratchetfreak** [comment]
I meant that the code appended for the function being "linked" in isn't a full module. Just the code from `OpFunction` (with a hardcoded result ID) up until `OpFunctionEnd` with the code using hardcoded type IDs and a range for intra function IDs.

Though looking at the other requirements for this to work it seems a much for glslang to do.

**Comment by Joojooo** [comment]
Thanks for your input. Our hlsl shaders are indeed partial and can define unused types or functions, then at the end we have a compilation system merging different hlsl shaders and only keeping the necessary attributes / functions. For now we parse and compile everything at the final stage but the parsing and compilation time can be very high, that's why we are checking to convert all those intermediary shaders into bytecode.

**Comment by johnkslang** [comment]
SPIR-V should be quite easy to manipulate to merge things properly, so for that side of the equation, I'd rather see investment in just doing it properly. But, it sounds like the issue is more on the front end.

So, you have one input compilation defining a type without using it, and then another compilation unit that uses the type without defining it? And it's not until they are all together that it makes sense? If so, you can certainly have custom tools to help you piece things together that way. E.g., add "@dummyVarN" to the linker objects to bring types into SPIR-V, and then when you see that in SPIR-V, keep the type but not the object. (The fundamental design is that the AST is about objects, and it will go easiest to stick to that.)

**Comment by Joojooo** [comment]
Yes that's correct, I can easily remove all unused types/variables when combining the SPIR-V files or before. So the question was more how to do it properly in the code.
The solution I'm using right now seems to work: after the parser create a type that I wanna force to be in the bytecode (whether it will be used or not), I call:

`
		TString unusedVariableName = "__var_" + declaredType.getTypeName() + "_unused";
		TString* pName = NewPoolTString(unusedVariableName.c_str());
		parseContext.declareVariable(initialToken.loc, *pName, declaredType, nullptr);
`
(The code in my previous post above wasn't calling NewPoolTString, sometimes making the app to crash)

Then I'll just need to remove this variable later on.

**Comment by johnkslang** [comment]
Another key point here is that glslang has an AST linker in it.  Glslang is already architected to take in multiple partial-stage compilation units for a single stage (single entry point). These partial stages are first independently compiled into partial ASTs, and then the linker part of glslang merges the multiple ASTs together (the whole "TIntermediate" actually) into a single one. The single one represents a complete call graph for an entry point, which can then be translated to valid SPIR-V, dropping anything not needed by the complete call graph rooted by the entry point.

The point of this goes hand in hand with not needing to have partial SPIR-V files merged to make a single stage; it would be handled by the front end, not by SPIR-V tools.  (The SPIR-V merging we do want is independent complete, not partial, SPIR-V modules merged into a single SPIR-V module with multiple entry points.)

So, it sounds like some of the trouble here comes from wanting to solve the "merge partial results" problem at the SPIR-V level instead of the AST level.

**Comment by dankbaker** [comment]
One thing I did notice is that in HLSL you can have static const literals evaluated from functions but GLSL/SPIRV doesn't' like this.

For example, we have some consts in our code:
static const float ODT_KNOT_START =  log10(rrt_tonescale_fwd(0.18));

which in the HLSL compiler auto resolves to:

static const float ODT_KNOT_START =  .698970f; 

But, this doesn't work in GLSLlang HLSL front end at least, I had to resolve them manually.

**Comment by johnkslang** [comment]
@dankbaker can you open an issue for this, and include rrt_tonescale_fwd()? Whatever the connection to passing in dead types to SPIR-V, I suspect a separate issue is best anyway, esp. given the extra information needed.
